File: static/cpp/java/src/main/java/com/epi/QueueWithMaxAlternative.java
package com.epi;

import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;
import java.util.NoSuchElementException;

public class QueueWithMaxAlternative {
  // @include
  public static class QueueWithMax<T extends Comparable<T>> {
    private Queue<T> entries = new LinkedList<>();
    private Deque<T> candidatesForMax = new LinkedList<>();

    public void enqueue(T x) {
      entries.add(x);
      while (!candidatesForMax.isEmpty()) {
        // Eliminate dominated elements in candidatesForMax.
        if (candidatesForMax.getLast().compareTo(x) >= 0) {
          break;
        }
        candidatesForMax.removeLast();
      }
      candidatesForMax.addLast(x);
    }

    public T dequeue() {
      if (!entries.isEmpty()) {
        T result = entries.remove();
        if (result.equals(candidatesForMax.getFirst())) {
          candidatesForMax.removeFirst();
        }
        return result;
      }
      throw new NoSuchElementException("Called dequeue() on empty queue.");
    }

    public T max() {
      if (!candidatesForMax.isEmpty()) {
        return candidatesForMax.getFirst();
      }
      throw new NoSuchElementException("empty queue");
    }
    // @exclude

    public T head() { return entries.peek(); }
    // @include
  }
  // @exclude

  private static <T extends Comparable<T>> void assertDequeue(QueueWithMax<T> q,
                                                              T t) {
    T dequeue = q.dequeue();
    assert(t.equals(dequeue));
  }

  private static void simpleTest() {
    QueueWithMax<Integer> Q = new QueueWithMax<>();
    Q.enqueue(11);
    Q.enqueue(2);
    assert(11 == Q.max());
    assertDequeue(Q, 11);
    assert(2 == Q.max());
    assertDequeue(Q, 2);
    Q.enqueue(3);
    assert(3 == Q.max());
    assertDequeue(Q, 3);
    Q.enqueue(Integer.MAX_VALUE - 1);
    Q.enqueue(Integer.MAX_VALUE);
    Q.enqueue(-2);
    Q.enqueue(-1);
    Q.enqueue(-1);
    Q.enqueue(Integer.MIN_VALUE);
    assert(Integer.MAX_VALUE == Q.max());
    assertDequeue(Q, Integer.MAX_VALUE - 1);
    assert(Integer.MAX_VALUE == Q.max());
    assertDequeue(Q, Integer.MAX_VALUE);
    assert(-1 == Q.max());
    assertDequeue(Q, -2);
    assert(-1 == Q.max());
    assertDequeue(Q, -1);
    assertDequeue(Q, -1);
    assert(Integer.MIN_VALUE == Q.max());
    assertDequeue(Q, Integer.MIN_VALUE);
    try {
      System.out.println("Q is empty, max() call should except = " + Q.max());
      assert(false);
    } catch (NoSuchElementException e) {
      System.out.println(e.getMessage());
    }
  }

  public static void main(String[] args) {
    simpleTest();
    QueueWithMax<Integer> Q = new QueueWithMax<>();
    Q.enqueue(1);
    Q.enqueue(2);
    assert(2 == Q.max());
    assertDequeue(Q, 1);
    assert(2 == Q.max());
    assertDequeue(Q, 2);
    Q.enqueue(3);
    assert(3 == Q.max());
    assertDequeue(Q, 3);
    try {
      Q.max();
      assert(false);
    } catch (NoSuchElementException e) {
      System.out.println(e.getMessage());
    }
    try {
      Q.dequeue();
      assert(false);
    } catch (NoSuchElementException e) {
      System.out.println(e.getMessage());
    }
  }
}

File: static/python/queue_with_max_alternative.py
import collections


# @include
class QueueWithMax:

    def __init__(self):
        self._entries = collections.deque()
        self._candidates_for_max = collections.deque()

    def enqueue(self, x):
        self._entries.append(x)
        # Eliminate dominated elements in _candidates_for_max.
        while self._candidates_for_max and self._candidates_for_max[-1] < x:
            self._candidates_for_max.pop()
        self._candidates_for_max.append(x)

    def dequeue(self):
        if self._entries:
            result = self._entries.popleft()
            if result == self._candidates_for_max[0]:
                self._candidates_for_max.popleft()
            return result
        raise IndexError('empty queue')

    def max(self):
        if self._candidates_for_max:
            return self._candidates_for_max[0]
        raise IndexError('empty queue')
# @exclude

    def head(self):
        return self._entries[0]


def simple_test():
    Q = QueueWithMax()
    Q.enqueue(11)
    Q.enqueue(2)
    assert 11 == Q.max()
    assert 11 == Q.dequeue()
    assert 2 == Q.max()
    assert 2 == Q.dequeue()
    Q.enqueue(3)
    assert 3 == Q.max()
    assert 3 == Q.dequeue()
    Q.enqueue(float('inf') - 1)
    Q.enqueue(float('inf'))
    Q.enqueue(-2)
    Q.enqueue(-1)
    Q.enqueue(-1)
    Q.enqueue(float('-inf'))
    assert float('inf') == Q.max()
    assert float('inf') - 1 == Q.dequeue()
    assert float('inf') == Q.max()
    assert float('inf') == Q.dequeue()
    assert -1 == Q.max()
    assert -2 == Q.dequeue()
    assert -1 == Q.max()
    assert -1 == Q.dequeue()
    assert -1 == Q.dequeue()
    assert float('-inf') == Q.max()
    assert float('-inf') == Q.dequeue()
    try:
        print('Q is empty, Max() call should except =', Q.max())
        assert False
    except IndexError as e:
        print(e)  # throw


def main():
    simple_test()
    Q = QueueWithMax()
    Q.enqueue(1)
    Q.enqueue(2)
    assert 2 == Q.max()
    assert 1 == Q.dequeue()  # 1
    assert 2 == Q.max()
    assert 2 == Q.dequeue()  # 2
    Q.enqueue(3)
    assert 3 == Q.max()
    assert 3 == Q.dequeue()  # 3
    try:
        Q.max()
        assert False
    except IndexError as e:
        print(e)  # throw
    try:
        Q.dequeue()
        assert True
    except IndexError as e:
        print(e)  # throw


if __name__ == '__main__':
    main()
